packages/langium/src/lsp/completion/completion-provider.ts

```diff
@@ -32,84 +32,123 @@ export interface CompletionProvider {
 }
 
 export class DefaultCompletionProvider implements CompletionProvider {
 
+    protected readonly completionParser: LangiumCompletionParser;
     protected readonly scopeProvider: ScopeProvider;
-    protected readonly ruleInterpreter: RuleInterpreter;
     protected readonly grammar: ast.Grammar;
+    protected readonly nameProvider: NameProvider;
 
     constructor(services: LangiumServices) {
         this.scopeProvider = services.references.ScopeProvider;
-        this.ruleInterpreter = services.lsp.completion.RuleInterpreter;
         this.grammar = services.Grammar;
+        this.completionParser = services.parser.CompletionParser;
+        this.nameProvider = services.references.NameProvider;
     }
 
-    getCompletion(document: LangiumDocument, params: CompletionParams): MaybePromise<CompletionList> {
+    getCompletion(document: LangiumDocument, params: CompletionParams): MaybePromise<CompletionList | undefined> {
         const root = document.parseResult.value;
         const cst = root.$cstNode;
-        const items: CompletionItem[] = [];
+        if (!cst) {
+            return undefined;
+        }
+        let items: CompletionItem[] = [];
         const offset = document.textDocument.offsetAt(params.position);
         const acceptor = (value: string | AstNode | AstNodeDescription, item?: Partial<CompletionItem>) => {
             const completionItem = this.fillCompletionItem(document.textDocument, offset, value, item);
             if (completionItem) {
                 items.push(completionItem);
             }
         };
-        if (cst) {
-            const node = findLeafNodeAtOffset(cst, offset);
-            if (node) {
-                const features = findNextFeatures(this.buildFeatureStack(node));
-                const commonSuperRule = this.findCommonSuperRule(node);
-                // In some cases, it is possible that we do not have a super rule
-                if (commonSuperRule) {
-                    const flattened = flattenCst(commonSuperRule.node).filter(e => e.offset < offset).toArray();
-                    const possibleFeatures = this.ruleInterpreter.interpretRule(commonSuperRule.rule, [...flattened], offset);
-                    // Remove features which we already identified during parsing
-                    const partialMatches = possibleFeatures.filter(e => {
-                        const match = this.ruleInterpreter.featureMatches(e, flattened[flattened.length - 1], offset);
-                        return match === 'partial' || match === 'both';
-                    });
-                    const notMatchingFeatures = possibleFeatures.filter(e => !partialMatches.includes(e));
-                    features.push(...partialMatches);
-                    features.push(...notMatchingFeatures.flatMap(e => findNextFeatures([e])));
+
+        const node = findLeafNodeAtOffset(cst, offset);
+
+        if (!node) {
+            const parserRule = getEntryRule(this.grammar)!;
```

Comment:
The case where `parserRule` is undefined could cause problems here with `!`.

Reply:
Since we know that the grammar in question is validated (it is only used after it has passed the validation), we know that there's always an entry rule.